### DISCUSSION TITLE
Docker install local plugins from volume

### DISCUSSION THREAD

**Discussion by luwol03** [inventree/InvenTree#3339] [Q&A] [answered]
Hello,

after trying out the demo instance for a while I decided to setup my own first installation via docker. Until now everything works fine. But now I want to add an own plugin but can't find some docs where I have to mount my plugin files. I'd be happy to see some simple example.

luwol03

**Comment by matmair** [comment, marked as answer]
Hi there @luwol03 you just need to install the plugin via the interface. It will be placed in the right place automatically. I am not sure what example you would expect - it is a single page form with 2 inputs.

**Reply by luwol03** [reply]
I cannot find that in the docs, where is that ui menu? I can only find a [documentation](https://inventree.readthedocs.io/en/latest/extend/plugins/#inventree-plugin-architecture) about adding it by placing it in a folder, but where is that in the ui? I don't want to use a plugin managed with a vcs, just a single self written plugin file like it is described with the mixings in the docs.

**Reply by SchrodingersGat** [reply]
@luwol03 are you trying to install a plugin with local source files, i.e. not installed via PIP?

If so, then I realize that this is currently not possible under docker. I've raised an issue to address this - https://github.com/inventree/InvenTree/issues/3343 - which will be available in the upcoming `0.8.0` release.

**Reply by luwol03** [reply]
@SchrodingersGat Yes, im trying to write and test my first plugin via a source file. Is there currently no work around for this, so that I don't have to wait for v0.8? I found [this](https://github.com/inventree/InvenTree/pull/2504) issue, but didn't understand how this should help me.

**Reply by SchrodingersGat** [reply]
Currently no work around, you are the first to try installing a plugin via local source under docker! 0.8.0 is not too far away, or you can try running from inventree:latest once the referenced patch is merged

**Reply by luwol03** [reply]
Ok. Thank you for tracking this.

**Reply by SchrodingersGat** [reply]
@luwol03 OK looking into this a bit more there should be a way to achieve this. I'd appreciate if you can test it for me...

I am making the following assumptions:

- You are running the docker stable production build, pulled down from dockerhub
- You have a local directory called `/my/local/plugin/directory/` (change this as necessary in the instructions)
- You are running a setup based on the suggested [docker compose script](https://github.com/inventree/InvenTree/blob/master/docker/production/docker-compose.yml)

From this starting point:

#### Add new external volume for plugins

At the end of your docker compose file, add something like this:

```yaml
    inventree_plugins:
        driver: local
        driver_opts:
          type: none
          o: bind
          device: /my/local/plugin/directory/
```

![image](https://user-images.githubusercontent.com/10080325/179398699-2a5cb2b1-4596-4fc4-880d-100d69e8cc41.png)

#### Mount volume to containers

Add this line (twice) to the `inventree-server` and `inventree-worker` containers:

![image](https://user-images.githubusercontent.com/10080325/179398740-5bc0d0cb-774c-43eb-8e9c-3ab19b51c945.png)

#### Add Plugins

Navigate on your local machine to `/my/local/plugin/directory/` and place your plugin files there. 

#### Restart Docker Containers

Run `docker-compose restart` to restart the system. This is required to re-initialize plugins.

**Reply by matmair** [reply]
@SchrodingersGat this seems like a good general solution.

Generally I would strongly suggest to use a VCS and release system for plugins you use in production at a bigger org. For personal deployments just the folder should be an easy and adequate solution.

**Reply by luwol03** [reply]
I was searching for something exactly like this. Thank you for that solution. I'll try that tomorrow and post if it worked.

Yes, VCS are Good, but it's hard to always pull the newest changes to test the plugin. I don't want to install it bare metal because I don't want to dirty my system with all kinds of dependencies and interpreters. (For that, docker is awesome)


### Open questions related
1. Would be good, if this could then be added to the docs if it's working, so that others can also profit from it.

2. And one other general question, why do you prefere defining volumes separately instead of a simple `./data:/path/in/container`?

**Reply by luwol03** [reply]
Yes, it worked. Can you please take a look to my questions 2? Im interested in an answer. 
And can you please add this to your initial comment so that we can mark it as an answer for this discussion?

**Reply by matmair** [reply]
@luwol03 please do not edit your answers except for fixing typos - we do not get notifications for that.
This is not a docker course - please look into their docs if you want to learn about the architecture. It is cleaner to define the volumes this way as there are many different drivers for docker and local is only one option of many.

We are not modifying answers generally - the whole thread gets highlighted and can be read to understand the whole solution probably.

**Reply by luwol03** [reply]
Ok, though only edited the headline to make it more clear, that these are open questions. Thank you for solving this. Custom plugins work now.
